src/NHibernate/Collection/Trackers/MapQueueOperationTracker.cs

```diff
@@ -0,0 +1,212 @@
+		{
+			return _removalQueue?.Contains(elementKey) ?? false;
+		}
+
+		/// <inheritdoc />
+		public override bool RemoveElementByKey(TKey elementKey, TValue oldElement, bool? existsInDb)
+		{
+			if (Cleared)
+			{
+				return _queue?.Remove(elementKey) ?? false;
+			}
+
+			// We can have the following scenarios:
+			// 1. remove a key that exists in db and it is not in the queue and removal queue (decrease queue size)
+			// 2. remove a key that exists in db and it is in the queue (decrease queue size)
+			// 3. remove a key that does not exist in db and it is not in the queue (don't decrease queue size)
+			// 4. remove a key that does not exist in db and it is in the queue (decrease queue size)
+			// 5. remove a key that exists in db and it is in the removal queue (don't decrease queue size)
+
+			// If the key is not present in the database and in the queue, do nothing
+			if (existsInDb == false && _queue?.ContainsKey(elementKey) != true)
+			{
+				return false;
+			}
+
+			if (existsInDb == true)
+			{
+				GetOrCreateOrphanMap()[elementKey] = oldElement;
+			}
+
+			// We don't want to have non database keys in the removal queue
+			if (_queue?.Remove(elementKey) == true |
+				(_orphanMap?.ContainsKey(elementKey) == true && GetOrCreateRemovalQueue().Add(elementKey)))
+			{
+				_queueSize--;
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <inheritdoc />
+		public override void SetElementByKey(TKey elementKey, TValue element, TValue oldElement, bool? existsInDb)
+		{
+			if (Cleared)
+			{
+				GetOrCreateQueue()[elementKey] = element;
+				return;
+			}
+
+			// We can have the following scenarios:
+			// 1. set a key that exists in db and it is not in the queue (don't increase queue size)
+			// 2. set a key that exists in db and it is in the queue (don't increase queue size)
+			// 3. set a key that does not exist in db and it is not in the queue (increase queue size)
+			// 4. set a key that does not exist in db and it is in the queue (don't increase queue size)
+			// 5. set a key that exists in db and it is in the removal queue (increase queue size)
+			if ((existsInDb == false && _queue?.ContainsKey(elementKey) != true) || _removalQueue?.Remove(elementKey) == true)
+			{
+				_queueSize++;
+			}
+
+			if (existsInDb == true)
+			{
+				GetOrCreateOrphanMap()[elementKey] = oldElement;
+			}
+
+			GetOrCreateQueue()[elementKey] = element;
+		}
+
+		/// <inheritdoc />
+		public override bool TryGetElementByKey(TKey elementKey, out TValue element)
+		{
+			if (_queue == null)
+			{
+				element = default(TValue);
+				return false;
+			}
+
+			return _queue.TryGetValue(elementKey, out element);
+		}
+
+		/// <inheritdoc />
+		public override bool TryGetDatabaseElementByKey(TKey elementKey, out TValue element)
+		{
+			if (_orphanMap == null)
+			{
+				element = default(TValue);
+				return false;
+			}
+
+			return _orphanMap.TryGetValue(elementKey, out element);
```

Comment:
In which case is-it useful to retrieve a persisted element from the to be orphaned ones? The only place where `TryGetDatabaseElementByKey` is currently called does previously check `IsElementKeyQueuedForDelete` and short-circuit in such case. So it looks to me this `TryGetDatabaseElementByKey` method is unneeded and could be removed.

Reply:
I didn't found any example where this would be used, so I removed it.